### PR TITLE
Add several configs to KafkaCruiseControlConfig and improve its usage in reflective created object.

### PR DESCRIPTION
**Pull request by kidkun**
Remove the config of pluggable component from centralized config definition in KafkaCruiseControlConfig so that user can have their own class implementation and customized config of these pluggable component and plug in to use with Cruise Control.

### CODE REVIEW

**Comment by becketqin** [comment]
@kidkun Thanks for the patch. I took a quick glance at the patch and have some concerns. It looks that the configurations moved to KafkaCruiseControlConfig are configurations for default implementation of the pluggbles (e.g. KafkaSampleStore). Personally I think it makes more sense to keep those configs in the specific implementation instead of putting them into the global configuration.

If we put those configurations into the global config, when user plug in another implementation, those global config will become unused and can easily cause confusions like "should I still set values for them". So IMO the global configuration should only have the configurations that are used by the framework itself regardless of the pluggins.

I know that currently we have some configurations there in the global config that only make sense to certain goals. It is a legacy issue and I actually think we should move them out to the corresponding goals.

**Comment by kidkun** [comment]
Hey Becket, thanks for the comment.  I find you suggestion make sense, actually I ignore the fact that some component are pluggable and put everything in CruiseControlConfig does harm our long goal of generalizing Cruise Control.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java [hunk 1]
```diff
@@ -55,17 +55,10 @@ public AbstractGoal() {
...
-    String minMonitoredPartitionPercentageString =
-        (String) configs.get(KafkaCruiseControlConfig.MIN_VALID_PARTITION_RATIO_CONFIG);
-    if (minMonitoredPartitionPercentageString != null
-        && !minMonitoredPartitionPercentageString.isEmpty()) {
-      _minMonitoredPartitionPercentage = Double.parseDouble(minMonitoredPartitionPercentageString);
-    }
+    KafkaCruiseControlConfig parsedConfig = new KafkaCruiseControlConfig(configs, false);
+    _balancingConstraint = new BalancingConstraint(parsedConfig);
+    _numWindows = parsedConfig.getInt(KafkaCruiseControlConfig.NUM_PARTITION_METRICS_WINDOWS_CONFIG);
+    _minMonitoredPartitionPercentage = parsedConfig.getDouble(KafkaCruiseControlConfig.MIN_VALID_PARTITION_RATIO_CONFIG);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java]: Do we still need to initialize `_numWindows` and `_minMonitoredPartitionPercentage` (see below) if they are guaranteed to be configured here?
```
  protected int _numWindows = 1;
  protected double _minMonitoredPartitionPercentage = 0.995;
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 2]
```diff
@@ -65,22 +65,24 @@
...
   protected String _brokerMetricSampleStoreTopic;
+  protected Integer _metricSampleStoreTopicReplicationFactor;
   protected volatile double _loadingProgress;
   protected Producer<byte[], byte[]> _producer;
   protected volatile boolean _shutdown = false;
 
+  public static final String PARTITION_METRIC_SAMPLE_STORE_TOPIC_CONFIG = "partition.metric.sample.store.topic";
+  public static final String BROKER_METRIC_SAMPLE_STORE_TOPIC_CONFIG = "broker.metric.sample.store.topic";
+  public static final String NUM_SAMPLE_LOADING_THREADS_CONFIG = "num.sample.loading.threads";
+  public static final String NUM_METRIC_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR_CONFIG = "num.metric.sample.store.topic.replication.factor";
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: Let's ensure that the functionality of each config copied to here is clear. Can we make it clear by adding the relevant config and its document as a JavaDoc for this class? -- e.g. for `NUM_SAMPLE_LOADING_THREADS_CONFIG` add `{@link #NUM_SAMPLE_LOADING_THREADS_CONFIG}: The config for the number of Kafka sample store consumer threads.` to the `JavaDoc` of this class.

Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: It looks like we lost the default values for the configs copied from `KafkaCruiseControlConfig.java`. Can we ensure that the default values are specified here as well? -- e.g. for `PARTITION_METRIC_SAMPLE_STORE_TOPIC_CONFIG`, the default value is "__KafkaCruiseControlPartitionMetricSamples".

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 3]
```diff
@@ -65,22 +65,24 @@
   protected ExecutorService _metricProcessorExecutor;
   protected String _partitionMetricSampleStoreTopic;
   protected String _brokerMetricSampleStoreTopic;
+  protected Integer _metricSampleStoreTopicReplicationFactor;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: `_metricSampleStoreTopicReplicationFactor` -> `_sampleStoreTopicReplicationFactor`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java [hunk 4]
```diff
@@ -21,17 +21,14 @@
...
-    _brokerMetricSampleStoreTopic = (String) config.get(KafkaCruiseControlConfig.BROKER_METRIC_SAMPLE_STORE_TOPIC_CONFIG);
-    if (_partitionMetricSampleStoreTopic == null
-        || _brokerMetricSampleStoreTopic == null
-        || _partitionMetricSampleStoreTopic.isEmpty()
-        || _brokerMetricSampleStoreTopic.isEmpty()) {
-      throw new IllegalArgumentException("The sample store topic names must be configured.");
-    }
-    String numProcessingThreadsString = (String) config.get(KafkaCruiseControlConfig.NUM_SAMPLE_LOADING_THREADS_CONFIG);
+    _partitionMetricSampleStoreTopic = KafkaCruiseControlUtils.getRequiredConfig(config, PARTITION_METRIC_SAMPLE_STORE_TOPIC_CONFIG);
+    _brokerMetricSampleStoreTopic = KafkaCruiseControlUtils.getRequiredConfig(config, BROKER_METRIC_SAMPLE_STORE_TOPIC_CONFIG);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java]: Should we ensure that if the user did not provide the config, we use the default value for it (as we used to do) -- e.g. `DEFAULT_BROKER_METRIC_SAMPLE_STORE_TOPIC`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java [hunk 5]
```diff
@@ -149,7 +151,8 @@ private void ensureTopicsCreated(Map<String, ?> config) {
         throw new IllegalStateException(String.format("Kafka cluster has no alive brokers. (zookeeper.connect = %s",
                                                       config.get(KafkaCruiseControlConfig.ZOOKEEPER_CONNECT_CONFIG)));
       }
-      int replicationFactor = Math.min(2, numberOfBrokersInCluster);
+      int replicationFactor = _metricSampleStoreTopicReplicationFactor == null ? Math.min(2, numberOfBrokersInCluster) :
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: Should we avoid hardcoding values in code? -- e.g. `2` --> can we move it to a variable `DEFAULT_...`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java (hunk 2 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: for `PARTITION_METRIC_SAMPLE_STORE_TOPIC_CONFIG` and `BROKER_METRIC_SAMPLE_STORE_TOPIC_CONFIG` we use `KafkaCruiseControlUtils.getRequiredConfig` which will throw exception if user does not explicitly specify the topic name while they are using this implementation of `SampleStore`. I feel this is reasonable. 

for `NUM_SAMPLE_LOADING_THREADS_CONFIG` we have the default value(set to 8). 

for `NUM_METRIC_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR_CONFIG`, it is set to `Math.min(2, numberOfBrokersInCluster)`.

I update the JavaDoc of this class to mention these are required configs. Do you think that is clear enough?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java (hunk 4 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java]: I think this is reasonable. If we auto create some topic that user are unaware of, they might be surprised to find out later and have some unwanted operation(like delete the topic...)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java (hunk 2 above)
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaSampleStore.java]: Sure, looks good to me!
-- let's make sure that we mention this behavior change in commit message.

**Pull request merged**